active-rfcs/0000-finalize-head-body.md

```diff
@@ -0,0 +1,80 @@
+- Start Date: 11/30/2021
+- Reference Issues: https://github.com/withastro/rfcs/discussions/15
+- Implementation PR:
+
+# Summary
+
+Finalize behavior around `<head>`, `<body>`, `<html>`, and `<!DOCTYPE html>` elements in Astro component templates.
+
+# Motivation
+
+in Astro v0.21, we have had user reports around `<head>` not acting as expected, breaking some projects from v0.20:
+- https://github.com/withastro/astro/issues/2046
+- https://github.com/withastro/astro/issues/2022
+
+Some of these issues stem from undocumented or undefined behaviors in v0.20: was it allowed for `<head>` be conditionally added in a nested expression? 
+
+Other issues stem from our attempt to provide implicit `<head>` and `<body>` support in pages and layouts. There was a quick workaround added for v0.21 where `src/pages` and `src/layouts` folder locations must be defined in config so that our parser knows how to parse them as full HTML documents and not smaller HTML fragments. This unblocked users but at the expense of breaking two larger design goals of the project:
+  - Added `src/layouts` as a new special folder location, when only `src/pages` was intented to be "special".
+  - Caused Astrp to create different output for a component based on the file path of that component.
+
+## Goals
+
+1. Agree on and document the rules and behaviors of the `<head>`, `<body>`, `<html>`, and `<!DOCTYPE html>` tags.
+2. Remove special compiler behavior based on where a file lives in your src directory. 
+3. Remove `src/layouts` as a "special" folder and any config around this.
+
+## Non-Goals
+
+1. **Out of scope:** Support `<head>` Injection inside of individual components. This is a commonly requested user feature, but can be tackled as an additional feature, leaving this RFC focused on clarifying our existing API. This RFC does not impact that feature request.
+
+
+# Detailed design
+
+## Template Changes
+
+- `<!DOCTYPE html>`
+  - compiler: completely ignored by the compiler, stripped out from render output.
+  - runtime: a `<!DOCTYPE html>` tag is always included in final page HTML output.
```

Comment:
why append this in the runtime rather than the compiler? makes more sense to me that the compiler generates essentially the exact HTML that will be sent to the user.

Reply:
The compiler is responsible for compiling every component individually. In this proposal, the compiler has no context for document (would need `<!DOCTYPE html>`) and component (would break if `<!DOCTYPE html>` were added). An Astro component is always considered a fragment of HTML.

Reply:
Ok, I follow that then.

Reply:
Does this mean adding `<!doctype html>` will definitely create a duplicate? I imagine the SSR answer is "yes".

Otherwise, you might need to scan the page, and then if the first thing is a `<Layout>` with `<!doctype html>`, doh!

Reply:
I read it as if you add a doctype as an author the compiler will remove it. Then when we go to render the page it just prepends the doctype always as the first bytes to go out.

Reply:
@matthewp is correct, that is the intention

Reply:
I understand what @matthewp wrote. Maybe that should be added?

I don’t agree with what you wrote, @FredKSchott, if you were saying a component should 'break' if it has a doctype.

```html
<!doctype html><!doctype html>
<style>.a { color: blue }</style>
<style>.b { font-weight: bold }</style><!doctype html>
<body class="a">
<body class="b">
<p>hello<!doctype html>
```

If the compiler does nothing, no harm. If the compiler omits it, no harm. If the compiler breaks? _Cruel_. 😂

Reply:
I mean, is it THAT cruel if our compiler can't handle that? :)

But yes, my intention in writing "compiler: completely ignored by the compiler, stripped out" is that it would ignored by the compiler and basically invisible, not included in the compiler render output. Let me know how I could clarify the RFC to make this more clear. This RFC would treat that code snippet as valid (although, arg!)

Reply:
Then I love it!

Reply:
🙌 🙌 🙌

Reply:
Is the reason for treating `<!doctype>` differently from, say, `<html>` spelled out somewhere? Sorry to drop in without context (thinking this has maybe been discussed elsewhere?), but would be great if there were a bit more explanation here.

Asking because, as a naive user it feels like the least surprising behavior would just be to output the doc as created in the templates I've defined, including the doctype and wrapping tags I've defined (or haven't defined, if I don't want those wrapping tags at all eg: https://github.com/withastro/astro/issues/1305#issuecomment-928219171 )

Reply:
> the least surprising behavior would just be to output the doc as created in the templates I've defined

To clarify: I’m open to preserving authored tags, too. I happen to like the convenience of `<!doctype html>`, but I wouldn’t stop using Astro if I had to add them myself. My only issue has been if Astro _injects_ tags that I never authored at all.

Reply:
I'm not strongly against changing this! This mainly comes from a place of making it easy to do the right thing, and hard to do the wrong thing. I think most developers are at a place where they don't even think about (or even know about) `<!doctype html>` and we should be supporting those users. I am almost positive they are in the majority of web developers these days.

So the main use-case that I want to support is: if I forgot to add `<!doctype html>`, Astro adds it for me. @jonathantneal @gabrielgrant if either of you want to propose a change to this item that you'd be happier with but still keeps that use-case/goal supported, I'm probably +1!

Reply:
Oh no, my clarification went sideways? Re: `<!doctype html>` I’m fine with injecting it, and I also wouldn’t be mad if we couldn’t for some reason. No strong opinion here. No plans to suggest alternatives. I like the convenience, and I agree with your sentiment about web developers, @FredKSchott.

Reply:
ah gotcha, thanks for the clarification @jonathantneal !

Reply:
I’m looking to build parts of a page with Astro. These parts are assembled at the edge, allowing them to have different caching behavior, not invalidating the edge cache due to only a change in the navigation for instance.
The automatic doctype makes use cases like this quite difficult. I guess the same would be true if you want to output an RSS feed or perhaps a dynamic SVG.